cvmfs/util/pipe.h

```diff
@@ -0,0 +1,271 @@
+  kPipeDetachedChild,
+  kPipeTest,
+  kPipeDownloadJobs,
+  kPipeDownloadJobsResults
+};
+
+/**
+ * Common signals used by pipes
+ */
+enum PipeSignals {
+  kPipeTerminateSignal = 1
+};
+
+template <PipeType pipeType>
+class CVMFS_EXPORT Pipe : public SingleCopy {
+  FRIEND_TEST(T_Util, ManagedExecRunShell);
+  FRIEND_TEST(T_Util, ManagedExecExecuteBinaryDoubleFork);
+  FRIEND_TEST(T_Util, ManagedExecExecuteBinaryAsChild);
+
+ public:
+  /**
+   * A pipe is a simple asynchronous communication mechanism. It establishes
+   * a unidirectional communication link between two file descriptors. One of
+   * them is used to only write to the pipe, the other one only to read from it.
+   *
+   * This class is a simple wrapper around the handling of a "standard" pipe
+   * that uses system calls.
+   *
+   * @note PipeType as class template parameter should symbolize the
+   *       functionality of the specific type, independent of what variable
+   *       name it has
+   */
+  Pipe() {
+    int pipe_fd[2];
+    MakePipe(pipe_fd);
+    fd_read_ = pipe_fd[0];
+    fd_write_ = pipe_fd[1];
+  }
+
+  /**
+   * Destructor closes all valid file descriptors of the pipe
+   */
+  ~Pipe() {
+    if (fd_read_ != -1) {
+      close(fd_read_);
+    }
+
+    if (fd_write_ != -1) {
+      close(fd_write_);
+    }
+  }
+
+  /**
+   * Closes all open file descriptors of the pipe and marks them as invalid
+   */
+  void Close() {
+    if (fd_read_ != -1) {
+      close(fd_read_);
+      fd_read_ = -1;
+    }
+
+    if (fd_write_ != -1) {
+      close(fd_write_);
+      fd_write_ = -1;
+    }
+  }
+
+  /**
+   * Closes file descriptor that reads from the pipe and marks it as invalid.
+   */
+  void CloseReadFd() {
+    close(fd_read_);
+    fd_read_ = -1;
+  }
+
+  /**
+   * Closes file descriptor that writes to the pipe and marks it as invalid.
+   */
+  void CloseWriteFd() {
+    close(fd_write_);
+    fd_write_ = -1;
+  }
+
+  /**
+   * Tries to write an object to the pipe
+   *
+   * @returns true if the entire object was written
+   *          false otherwise
+   */
+  template<typename T>
+  bool TryWrite(const T &data) {
+    const int num_bytes = write(fd_write_, &data, sizeof(T));
+    return (num_bytes >= 0) && (static_cast<size_t>(num_bytes) == sizeof(T));
+  }
+
+  /**
+   * Writes an object to the pipe
+   *
+   * @returns true on success
+   *          otherwise kills the program with an assert
+   *
+   */
+  template<typename T>
+  bool Write(const T &data) {
+    WritePipe(fd_write_, &data, sizeof(T));
+    return true;
+  }
+
+  /**
+   * Writes an object to the pipe
+   * If possible, it is recommended to use "bool Write(const T &data)"
+   *
+   * @returns true on success
+   *          otherwise kills the program with an assert
+   *
+   */
+  bool Write(const void *buf, size_t nbyte) {
+    WritePipe(fd_write_, buf, nbyte);
+    return true;
+  }
+
+  /**
+   * Tries to read from the pipe until it receives data or it is interrupted
```

Comment:
It is actually not breaking when interrupted but retries to read. The point is that it accepts partial reads without panicking.

Reply:
changed comment to 
```
   * (Re)tries to read from the pipe until it receives data or returned error
   * is NOT a system interrupt
  ```
  
might this be related to the weird double write of data dave noticed?

Reply:
I'd be surprised because it only re-reads if it didn't receive any data yet.